File: code/richediting/src/main/java/org/fbme/ide/richediting/inspections/ECCInspectionsFacility.java
package org.fbme.ide.richediting.inspections;

import org.fbme.ide.richediting.adapters.ecc.ECTransitionController;
import org.fbme.lib.iec61499.ecc.ECC;
import org.fbme.lib.iec61499.ecc.StateTransition;
import org.fbme.scenes.controllers.scene.SceneEditor;
import org.fbme.scenes.controllers.scene.SceneInitializer;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.awt.*;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.function.Function;

public class ECCInspectionsFacility {

    @NotNull
    private final ECC myECC;

    @NotNull
    private final SceneEditor myEditor;

    @NotNull
    private final Function<StateTransition, ECTransitionController> myTransitionProvider;

    private final Set<StateTransition> myHighlightedTransitions = new HashSet<>();

    public ECCInspectionsFacility(@NotNull ECC ecc, @NotNull SceneEditor editor, @NotNull Function<StateTransition, ECTransitionController> connectionProvider) {
        myECC = ecc;
        myEditor = editor;
        myTransitionProvider = connectionProvider;
        myEditor.addInitializer(new MyInitializer());
    }

    public void setHighlightingForTransition(@NotNull StateTransition transition, @Nullable Color color) {
        ECTransitionController controller = myTransitionProvider.apply(transition);
        if (color != null) {
            myHighlightedTransitions.add(transition);
        } else {
            myHighlightedTransitions.remove(transition);
        }
        controller.setHighlightColor(color);
        myEditor.fireRepaint();
    }

    public void clear() {
        for (StateTransition transition : myHighlightedTransitions) {
            ECTransitionController controller = myTransitionProvider.apply(transition);
            controller.setHighlightColor(null);
        }
        myHighlightedTransitions.clear();
        myEditor.fireRepaint();
    }

    private class MyInitializer implements SceneInitializer {

        @Override
        public void onAdd() {
            ECCInspectionsData inspectionsData = myEditor.loadState(ECCInspectionsData.KEY);
            if (inspectionsData != null) {
                loadState(inspectionsData);
            }
            InspectionManagerImpl.getInstance(myEditor.getContext().getEditorComponent()).registerECC(myECC, ECCInspectionsFacility.this);
        }

        @Override
        public void onRemove() {
            ECCInspectionsData data = new ECCInspectionsData();
            storeState(data);
            myEditor.storeState(ECCInspectionsData.KEY, data);
            InspectionManager manager = InspectionManagerImpl.getInstance(myEditor.getContext().getEditorComponent());
            if (manager != null) {
                manager.unregisterECC(myECC);
            }
        }
    }

    private void loadState(ECCInspectionsData data) {
        for (Map.Entry<StateTransition, Color> transition : data.highlightedTransitions.entrySet()) {
            myHighlightedTransitions.add(transition.getKey());
            ECTransitionController controller = myTransitionProvider.apply(transition.getKey());
            controller.setHighlightColor(transition.getValue());
        }
        if (!data.highlightedTransitions.isEmpty()) {
            myEditor.fireRepaint();
        }
    }

    private void storeState(ECCInspectionsData data) {
        for (StateTransition connection : myHighlightedTransitions) {
            ECTransitionController controller = myTransitionProvider.apply(connection);
            data.highlightedTransitions.put(connection, controller.getHighlightColor());
        }
    }

}

File: code/richediting/src/main/kotlin/org/fbme/ide/richediting/inspections/ECCInspectionsFacility.kt
package org.fbme.ide.richediting.inspections

import org.fbme.ide.richediting.adapters.ecc.ECTransitionController
import org.fbme.lib.iec61499.ecc.ECC
import org.fbme.lib.iec61499.ecc.StateTransition
import org.fbme.scenes.controllers.scene.SceneEditor
import org.fbme.scenes.controllers.scene.SceneInitializer
import java.awt.Color
import java.util.function.Function

class ECCInspectionsFacility(
    private val ecc: ECC,
    private val editor: SceneEditor,
    private val transitionProvider: Function<StateTransition, ECTransitionController>
) {
    private val highlightedTransitions: MutableSet<StateTransition> = HashSet()
    fun setHighlightingForTransition(transition: StateTransition, color: Color?) {
        val controller = transitionProvider.apply(transition)
        if (color != null) {
            highlightedTransitions.add(transition)
        } else {
            highlightedTransitions.remove(transition)
        }
        controller.highlightColor = color
        editor.fireRepaint()
    }

    fun clear() {
        for (transition in highlightedTransitions) {
            val controller = transitionProvider.apply(transition)
            controller.highlightColor = null
        }
        highlightedTransitions.clear()
        editor.fireRepaint()
    }

    private inner class MyInitializer : SceneInitializer {
        override fun onAdd() {
            val inspectionsData = editor.loadState(
                ECCInspectionsData.KEY
            )
            inspectionsData?.let { loadState(it) }
            val manager = InspectionManagerImpl.getInstance(editor.editorContext.editorComponent)
            manager?.registerECC(ecc, this@ECCInspectionsFacility)
        }

        override fun onRemove() {
            val data = ECCInspectionsData()
            storeState(data)
            editor.storeState(ECCInspectionsData.KEY, data)
            val manager = InspectionManagerImpl.getInstance(editor.editorContext.editorComponent)
            manager?.unregisterECC(ecc)
        }
    }

    private fun loadState(data: ECCInspectionsData) {
        for ((key, value) in data.highlightedTransitions) {
            highlightedTransitions.add(key)
            val controller = transitionProvider.apply(key)
            controller.highlightColor = value
        }
        if (data.highlightedTransitions.isNotEmpty()) {
            editor.fireRepaint()
        }
    }

    private fun storeState(data: ECCInspectionsData) {
        for (connection in highlightedTransitions) {
            val controller = transitionProvider.apply(connection)
            data.highlightedTransitions[connection] = controller.highlightColor
        }
    }

    init {
        editor.addInitializer(MyInitializer())
    }
}
